src/server/page_server.cpp

```diff
@@ -107,64 +109,151 @@ page_server::connection_handler::receive_log_prior_list (tran_server_conn_t::seq
+{
+  push_async_response (&connection_handler::fetch_log_page, std::move (a_sp));
+}
+
+void
+page_server::connection_handler::fetch_data_page (cubthread::entry &context, std::string &a_payload)
 {
-  std::string message = a_sp.pull_payload ();
+  // Unpack the message data
+  cubpacking::unpacker message_upk (a_payload.c_str (), a_payload.size ());
 
-  cubpacking::unpacker message_upk (message.c_str (), message.size ());
   VPID vpid;
   vpid_utils::unpack (message_upk, vpid);
 
   LOG_LSA target_repl_lsa;
   cublog::lsa_utils::unpack (message_upk, target_repl_lsa);
 
-  auto callback_func = [this, sp = a_sp] (const FILEIO_PAGE *iopage, int error_code) mutable
-  {
-    on_data_page_read_result (std::move (sp), iopage, error_code);
-  };
+  // Fetch data page. But first make sure that replication hits its target LSA
+  if (!target_repl_lsa.is_null ())
+    {
+      // TODO: FIXME
+      // The transaction server boots and reads pages before initializing its log module and before knowing a safe target
+      // LSA for replication. A way of knowing this target LSA is required, but disable this wait until that's fixed.
+      m_ps.get_replicator ().wait_past_target_lsa (target_repl_lsa);
+    }
+
+  int error = NO_ERROR;
+  PAGE_PTR page_ptr = pgbuf_fix (&context, &vpid, OLD_PAGE, PGBUF_LATCH_READ, PGBUF_UNCONDITIONAL_LATCH);
+  if (page_ptr == nullptr)
+    {
+      ASSERT_ERROR_AND_SET (error);
+      // respond with the error
+      a_payload = { reinterpret_cast<const char *> (&error), sizeof (error) };
 
-  m_ps.get_page_fetcher ().fetch_data_page (vpid, target_repl_lsa, std::move (callback_func));
+      if (prm_get_bool_value (PRM_ID_ER_LOG_READ_DATA_PAGE))
+	{
+	  _er_log_debug (ARG_FILE_LINE,
+			 "[READ DATA] Error %lld while fixing page %d|%d, replication LSA = %lld|%d\n",
+			 error, VPID_AS_ARGS (&vpid), LSA_AS_ARGS (&target_repl_lsa));
+	}
+    }
+  else
+    {
+      FILEIO_PAGE *io_pgptr = nullptr;
+      pgbuf_cast_pgptr_to_iopgptr (page_ptr, io_pgptr);
+      assert (io_pgptr != nullptr);
+
+      // pack NO_ERROR first
+      a_payload = { reinterpret_cast<const char *> (&error), sizeof (error) };
+
+      // add io_page
+      a_payload.append (reinterpret_cast<const char *> (io_pgptr), (size_t) db_io_page_size ());
+
+      pgbuf_unfix (&context, page_ptr);
+
+      if (prm_get_bool_value (PRM_ID_ER_LOG_READ_DATA_PAGE))
+	{
+	  _er_log_debug (ARG_FILE_LINE,
+			 "[READ DATA] Successful while fixing page %d|%d, replication LSA = %lld|%d\n",
+			 VPID_AS_ARGS (&vpid), LSA_AS_ARGS (&target_repl_lsa));
+	}
+    }
 }
 
 void
-page_server::connection_handler::receive_log_boot_info_fetch (tran_server_conn_t::sequenced_payload &a_sp)
+page_server::connection_handler::receive_data_page_fetch (tran_server_conn_t::sequenced_payload &a_sp)
 {
-  // empty request message
+  push_async_response (&connection_handler::fetch_data_page, std::move (a_sp));
```

Comment:
this is where `push_async_response` must be added a new flavor for non-member handlers